gobblin-data-management/src/main/java/org/apache/gobblin/data/management/source/LoopingDatasetFinderSource.java

```diff
@@ -85,31 +86,52 @@ public WorkUnitStream getWorkunitStream(SourceState state) {
       int maxWorkUnits = state.getPropAsInt(MAX_WORK_UNITS_PER_RUN_KEY, MAX_WORK_UNITS_PER_RUN);
 
       List<WorkUnitState> previousWorkUnitStates = state.getPreviousWorkUnitStates();
-      Optional<WorkUnitState> maxWorkUnit;
-      try {
-        maxWorkUnit = previousWorkUnitStates.stream().reduce((wu1, wu2) -> {
-          int wu1Ordinal = wu1.getPropAsInt(WORK_UNIT_ORDINAL);
-          int wu2Ordinal = wu2.getPropAsInt(WORK_UNIT_ORDINAL);
-          return wu1Ordinal > wu2Ordinal ? wu1 : wu2;
-        });
-      } catch (NumberFormatException nfe) {
-        throw new RuntimeException("Work units in state store are corrupted! Missing or malformed " + WORK_UNIT_ORDINAL);
+      int tempWorkUnitOrdinal = 0;
+      Optional<WorkUnitState> optionalMaxWorkUnit = Optional.empty();
+      List<WorkUnit> failedPreviousWorkUnits = new ArrayList<>();
+      int maxRetryTimes = state.contains(ConfigurationKeys.MAX_TASK_RETRIES_KEY) ? state
+          .getPropAsInt(ConfigurationKeys.MAX_TASK_RETRIES_KEY) : ConfigurationKeys.DEFAULT_MAX_TASK_RETRIES;
+
+      // Searching for workunit with maximum ordinal meanwhile collect all failed workunit in previous run.
+      // These two processes should not influent each other.
+
+      // Workunits that exceed max retry limit will not be added into failedPreviousWorkUnits
+      // and corresponding event/notification should be sent in the task level.
```

Comment:
You are not sending events right now, are you? Also, instead of just discarding it and sending an event, we should execute the work unit on the next run.